### ISSUE TITLE
Optimizing the RequestIdentifier

### ISSUE THREAD

**Issue by morrys** [facebook/relay#2984]
Now the RequestIdentifier is created by concatenating the id or text of RequestParameters with the serialization of the variables.

https://github.com/facebook/relay/blob/master/packages/relay-runtime/util/getRequestIdentifier.js#L27-L39
```
function getRequestIdentifier(
  parameters: RequestParameters,
  variables: Variables,
): RequestIdentifier {
  const requestID = parameters.id != null ? parameters.id : parameters.text;
  invariant(
    requestID != null,
    'getRequestIdentifier: Expected request `%s` to have either a ' +
      'valid `id` or `text` property',
    parameters.name,
  );
  return requestID + JSON.stringify(stableCopy(variables));
}
```

When there is no id, the RequestIdentifier is a very long string (all the text of the query).

Wouldn't it be better to use the query hash value?

Let me know and I'll proceed with creating the PR.

**Comment by kassens** [comment]
Good catch, I've noticed this before but didn't get around to fixing this.

I think it'd make sense to add a `hash` field in addition to `text` to the query parameters object and use here (if we don't have an `id`).

**Comment by kassens** [comment]
Closed by #2985
